### DISCUSSION TITLE
What do you think about migrating to Pigeon?

### DISCUSSION THREAD

**Discussion by ycherniavskyi** [flutter-webrtc/flutter-webrtc#1011] [Ideas]
I have migrated my in-house plugin to [Pigeon](https://pub.dev/packages/pigeon) from direct platform channels manipulation, and I am very satisfied with the result.

So I would like to try to migrate [flutter-webrtc](https://github.com/flutter-webrtc/flutter-webrtc) to it.

But I need to know any objections about such migration from the community. Because PR will be huge, and if it is not going to applied, I will not start it.

**Comment by cloudwebrtc** [comment]
Looks like a quick way to build a plugin API, I'm curious if it's through ffi or the flutter method channel, I read pigeon's readme and it looks like it can drastically reduce hand-coding.

This seems to be a code generator
https://github.com/flutter/packages/tree/main/packages/pigeon/example#invocation

@ycherniavskyi Does pigeon have any implementations for handling event channels and texture rendering?

**Reply by ycherniavskyi** [reply]
Yes, exactly. It is a code generator that, from dart classes/methods marked with specific annotations (here are examples of such [pigeons files](https://github.com/flutter/packages/tree/main/packages/pigeon/pigeons)), generates all necessary glue code via flutter channel.

It is support **method channel** via `HostApi()` annotation and **event channel** via  `FlutterApi()`, but internal it use **basic message channel**. What was a bit odd for me that it create separate channel for each method 🤔, but when I see that flutter team migrate to it, I think that they know what they are doing 🙂.

As for texture rendering, as I see it, it does not provide any special helpers. And I think that I could start from `RTCVideoRenderer` to make a small demo of [Pigeon](https://pub.dev/packages/pigeon).

As for using `ffi`, for me, it will be even better, but does the WebRTC library provide a general C/C++ API for all platforms?

**Reply by ycherniavskyi** [reply]
I need to point out that [Pigeon](https://pub.dev/packages/pigeon) file has a limited set of Dart language functionality. That is why interfaces/classes in this file could not be used as public. They must be perceived as "type-safe customized flutter channels".

**Comment by ycherniavskyi** [comment]
Also, as an additional argument for such migration, I could say that flutter team in the process of migrating their [plugins](https://github.com/flutter/plugins) to [Pigeon](https://pub.dev/packages/pigeon) as well.

**Reply by cloudwebrtc** [reply]
@davidliu @hiroshihorie @davidzhao  I did some research about Pigeon, it should be a better way to encapsulate the API between flutter and native, it should be able to greatly reduce the process of manually writing method invoke and event channel, reduce errors, and even convert existing native code Migrating from oc to swift, from java to Kotlin

Message and method definitions
```dart
// pigeons/message.dart

import 'package:pigeon/pigeon.dart';

class F2NMessage{
  String? msg;
}

class N2FMessage{
  String? msg2;
}

@HostApi()
abstract class FlutterMessage{
  void flutterSendMessage(F2NMessage msg);
}

@FlutterApi()
abstract class NativeMessage{
  void nativeSendMessage(N2FMessage msg);
}
```

generate code
```bash
mkdir -p android/src/main/java/com/example/flutter_pigeon_plugin

flutter pub run pigeon \
  --input pigeons/message.dart \
  --dart_out lib/message.dart \
  --objc_header_out ios/Classes/Message.h \
  --objc_source_out ios/Classes/Message.m \
  --java_out android/src/main/java/com/example/flutter_pigeon_plugin/Message.java \
  --java_package "com.example.flutter_pigeon_plugin"
```

Flutter plugin implementation
```dart
// lib/flutter_pigeon_plugin.dart

import 'package:flutter_pigeon_plugin/message.dart';

class FlutterPigeonPlugin extends NativeMessage {
  static final FlutterMessage _flutterMessage = FlutterMessage();

  factory FlutterPigeonPlugin() => _getInstance();

  static FlutterPigeonPlugin get instance => _getInstance();

  static FlutterPigeonPlugin? _instance = null;

  static FlutterPigeonPlugin _getInstance() {
    _instance ??= FlutterPigeonPlugin._internal();
    NativeMessage.setup(_instance);
    return _instance!;
  }

  FlutterPigeonPlugin._internal();

  void sendMessage() {
    F2NMessage message = F2NMessage();
    message.msg = "This is a message sent from Flutter to Native";
    _flutterMessage.flutterSendMessage(message);
  }

  @override
  void nativeSendMessage(N2FMessage msg) {
    print("Received a message from Native：${msg.msg2}");
  }
}
```


swift native implementation
```swift
// ios/Classes/SwiftFlutterPigeonPlugin.swift


import Flutter
import UIKit

public class SwiftFlutterPigeonPlugin: NSObject, FlutterPlugin,FlutterMessage {
    public func flutterSendMsg(_ msg: F2NMessage, error: AutoreleasingUnsafeMutablePointer<FlutterError?>) {
        print("Receive a message from Flutter:"+msg.msg!)
        let message:N2FMessage = N2FMessage.make(withMsg2: "message from iOS send to Flutter")
        nativeMessage.nativeSendMsg(message, completion: {e in})
    }
    public var nativeMessage:NativeMessage!
    public static func register(with registrar: FlutterPluginRegistrar) {
        let messenger: FlutterBinaryMessenger = registrar.messenger()
        let api = SwiftFlutterPigeonPlugin.init()
        FlutterMessageSetup(messenger, api)
        api.nativeMessage = NativeMessage(binaryMessenger: messenger)
    }
}

```

kt native implementation
```kt
// src/main/kotlin/com/example/flutter_pigeon_plugin/FlutterPigeonPlugin.kt

package com.example.flutter_pigeon_plugin

import android.util.Log
import androidx.annotation.NonNull

import io.flutter.embedding.engine.plugins.FlutterPlugin

/** FlutterPigeonPlugin */
class FlutterPigeonPlugin : FlutterPlugin, Message.FlutterMessage {
    lateinit var nativeMessage: Message.NativeMessage

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        Message.FlutterMessage.setup(flutterPluginBinding.binaryMessenger, this)
        nativeMessage = Message.NativeMessage(flutterPluginBinding.binaryMessenger)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {

    }

    override fun flutterSendMessage(msg: Message.F2NMessage) {
        Log.d("FlutterPigeonPlugin", msg.msg.toString())
        var respMsg: Message.N2FMessage = Message.N2FMessage()
        respMsg.msg2 = "message from Android send to Flutter"
        nativeMessage.nativeSendMessage(respMsg, {})

    }
}
```

**Comment by hiroshihorie** [comment]
type-safety sounds great since we had bugs related to this.
(different platform returns different types for DataChannel.id etc)

but sounds like a lot of work ?

**Reply by ycherniavskyi** [reply]
Yes, it will not be effortless, but it will build a solid background for future extending 😎.

**Comment by davidzhao** [comment]
sounds excellent in terms of being an improvement to robustness. main question is wondering if there's a way to incrementally convert functions instead of having to do it in a single giant PR.

**Reply by ycherniavskyi** [reply]
Yes, for sure, at last, the flutter team move exactly this way.

So by the end of this week, I am going to migrate some Flutter RTC class(es) to [Pigeon](https://pub.dev/packages/pigeon), so we can make our final decision.

**Reply by cloudwebrtc** [reply]
Sounds great, it looks like C++ is supported，
I did some research and it seems to be perfect except the host API for flutter_linux/glibc cannot be generated

But I wrote a bridge code from flutter_linux to flutter C++ before
https://github.com/flutter-webrtc/flutter-webrtc/tree/main/linux/flutter
